src/main/java/build/buildfarm/instance/shard/ShardInstance.java

```diff
@@ -193,15 +196,14 @@
   private final com.google.common.cache.LoadingCache<String, Instance> workerStubs;
   private final Thread dispatchedMonitor;
   private final Duration maxActionTimeout;
-  private final Cache<Digest, Directory> directoryCache =
-      CacheBuilder.newBuilder().maximumSize(64 * 1024).build();
-  private final Cache<Digest, Command> commandCache =
-      CacheBuilder.newBuilder().maximumSize(64 * 1024).build();
-  private final Cache<Digest, Action> actionCache =
-      CacheBuilder.newBuilder().maximumSize(64 * 1024).build();
-  private final com.google.common.cache.Cache<RequestMetadata, Boolean>
-      recentCacheServedExecutions =
-          com.google.common.cache.CacheBuilder.newBuilder().maximumSize(64 * 1024).build();
+  private final AsyncCache<Digest, Directory> directoryCache =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
+  private final AsyncCache<Digest, Command> commandCache =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
+  private final AsyncCache<Digest, Action> actionCache =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
+  private final AsyncCache<RequestMetadata, Boolean> recentCacheServedExecutions =
+      Caffeine.newBuilder().newBuilder().maximumSize(64 * 1024).buildAsync();
```

Comment:
This should not be an async cache - note the use of the explicit guava upstream synchronous cache type. If you want to use caffeine for everything (I suppose I'm fine with that) then at least use the sync versions of their caches.

I also don't know how you got away with this without changing any of the usages, unless they default to synchronous, and we're only using the sync api.

Reply:
fixed. back to regular Cache